src/bson.c

```diff
@@ -1330,17 +1280,16 @@ void phongo_zval_to_bson(zval *data, php_phongo_bson_flags_t flags, bson_t *bson
 
 		if (hash_type == HASH_KEY_IS_STRING) {
 			if (ht_data_from_properties) {
-				const char *class_name;
-				zend_unmangle_property_name(string_key, string_key_len-1, &class_name, (const char **)&string_key);
-				string_key_len = strlen(string_key);
-
-				/* Ignore non-public properties */
-				if (!is_public_property(Z_OBJCE_P(data), string_key, string_key_len TSRMLS_CC)) {
+				/* Skip protected and private properties */
+				if (string_key[0] == '\0' && string_key_len > 1) {
 					continue;
 				}
-			} else {
-				/* Chop off the \0 from string lengths */
-				string_key_len -= 1;
+			}
+
+			if (strlen(string_key) != string_key_len - 1) {
+				phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "BSON keys cannot contain null bytes. Unexpected null byte after \"%s\".", ZSTR_VAL(string_key));
```

Comment:
Added this exception message. This appears after we might skip object properties whose name contains a leading null byte (i.e. PHP's encoding for protected and private members). For arrays, the check happens on all string keys.

We don't check this for integer keys, since we convert those to strings ourself and don't have to worry about null bytes.